src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -767,14 +878,160 @@ private static ISymbol SymbolFromArrayCreationOperation(IArrayCreationOperation
+            if (constructorArguments.Length != 0)
             {
-                if (symbol.HasAttribute(previewFeatureAttribute))
+                if (constructorArguments.First().Value is string messageValue)
                 {
-                    requiresPreviewFeaturesSymbols.GetOrAdd(symbol, true);
-                    return true;
+                    message = messageValue;
+                }
+            }
+
+            ImmutableArray<System.Collections.Generic.KeyValuePair<string, TypedConstant>> namedArguments = attribute.NamedArguments;
+            if (namedArguments.Length != 0)
+            {
+                foreach (System.Collections.Generic.KeyValuePair<string, TypedConstant> namedArgument in namedArguments)
+                {
+                    if (namedArgument.Key == "Message")
+                    {
+                        message = (string)namedArgument.Value.Value;
+                    }
+
+                    if (namedArgument.Key == "Url")
+                    {
+                        url = (string)namedArgument.Value.Value;
+                    }
+                }
+            }
+
+            return message;
+        }
+
+        private static void ReportDiagnosticWithCustomMessageIfItExists(OperationAnalysisContext context,
+                                                                        IOperation operation,
+                                                                        ISymbol symbol,
+                                                                        ConcurrentDictionary<ISymbol, ValueTuple<bool, string?, string?>> requiresPreviewFeaturesSymbols,
+                                                                        DiagnosticDescriptor diagnosticDescriptor,
+                                                                        DiagnosticDescriptor diagnosticDescriptorWithPlaceholdersForCustomMessage,
+                                                                        string diagnosticMessageArgument)
+        {
+            if (!requiresPreviewFeaturesSymbols.TryGetValue(symbol, out ValueTuple<bool, string?, string?> existing))
+            {
+                Debug.Assert(true, $"Should never reach this line. This means the symbol {symbol.Name} was not processed in this analyzer");
+            }
+            else
+            {
+                string url = existing.Item3 ?? DefaultURL;
+                if (existing.Item2 is string customMessage)
+                {
+                    context.ReportDiagnostic(operation.CreateDiagnostic(diagnosticDescriptorWithPlaceholdersForCustomMessage, diagnosticMessageArgument, url, customMessage));
+                }
+                else
+                {
+                    context.ReportDiagnostic(operation.CreateDiagnostic(diagnosticDescriptor, diagnosticMessageArgument, url));
+                }
+            }
+        }
+
+        private static void ReportDiagnosticWithCustomMessageIfItExists(SymbolAnalysisContext context,
```

Comment:
NIT: these overloads implementations are so similar, some parts might could be reused

Reply:
agree, just make all the diagnosticMessageArgument0..X optional parameters and check if they are null in a single methid instead of us having to maintain identical copies

Reply:
Ah yea each method is just different enough that I made a new method for it :) Will clean it up in the final pass if there's an intuitive way to refactor the methods